rpcs3/Emu/RSX/Overlays/overlay_osk.cpp

```diff
@@ -29,37 +32,111 @@ namespace rsx
 			fade_animation.active = true;
 		}
 
-		void osk_dialog::initialize_layout(const std::vector<grid_entry_ctor>& layout, const std::u32string& title, const std::u32string& initial_text)
+		void osk_dialog::add_panel(const osk_panel& panel)
+		{
+			// On PS3 apparently only 7 panels are added, the rest is ignored
+			if (m_panels.size() < 7)
+			{
+				// Don't add this panel if there already exists one with the same panel mode
+				for (const auto& existing : m_panels)
+				{
+					if (existing.osk_panel_mode == panel.osk_panel_mode)
+					{
+						return;
+					}
+				}
+				m_panels.push_back(panel);
+			}
+		}
+
+		void osk_dialog::step_panel(bool next_panel)
+		{
+			const size_t num_panels = m_panels.size();
+
+			if (num_panels > 0)
+			{
+				if (next_panel)
+				{
+					m_panel_index = (m_panel_index + 1) % num_panels;
+				}
+				else if (m_panel_index > 0)
+				{
+					m_panel_index = (m_panel_index - 1) % num_panels;
+				}
+				else
+				{
+					m_panel_index = num_panels - 1;
+				}
+			}
+
+			update_panel();
+		}
+
+		void osk_dialog::update_panel()
 		{
+			ASSERT(m_panel_index < m_panels.size());
+
+			const auto panel = m_panels[m_panel_index];
```

Comment:
why not `const auto&`? The compiler inserts it for you but still.

Reply:
Nope, this is a copy.